src/main/java/org/reaktivity/nukleus/http_cache/internal/stream/HttpProxyCacheableRequestGroup.java

```diff
@@ -60,88 +87,218 @@ void setRecentAuthorizationToken(
+                                INITIATE_REQUEST_SIGNAL);
+            }
         }
     }
 
-    void onNonCacheableResponse(
-        long acceptReplyId)
+    void onCacheableResponseUpdated(
+        String etag)
     {
-        routeIdsByReplyId.remove(acceptReplyId);
-        serveNextRequestIfPossible();
+        requestsQueue.forEach((key, routeIdsByReplyId) ->
+        {
+            if (key != null && key.equals(etag) ||
+                (key == null && etag == null) && requestsQueue.size() > 1)
+            {
+                routeIdsByReplyId.forEach(this::doSignalCacheEntryNotModified);
+            }
+            else
+            {
+                routeIdsByReplyId.forEach(this::doSignalCacheEntryUpdated);
+            }
+        });
     }
 
-    void onCacheableResponseUpdated()
+    void onCacheableResponseAborted()
     {
-        routeIdsByReplyId.forEach(this::doSignalCacheEntryUpdated);
+        requestsQueue.forEach((key, routeIdsByReplyId) ->
+        {
+            routeIdsByReplyId.forEach(this::doSignalCacheEntryAborted);
+        });
     }
 
-    void onCacheableResponseAborted()
+    void onGroupRequestReset()
+    {
+        factory.writer.doSignal(acceptRouteId,
+                                acceptReplyId,
+                                factory.supplyTraceId.getAsLong(),
+                                GROUP_REQUEST_RESET_SIGNAL);
+        acceptRouteId = 0L;
+        acceptReplyId = 0L;
+    }
+
+    MessageConsumer newRequest(
+        int msgTypeId,
+        DirectBuffer buffer,
+        int index,
+        int length,
+        MessageConsumer sender)
     {
-        this.routeIdsByReplyId.forEach(this::doSignalCacheEntryAborted);
+        BeginFW begin = factory.beginRO.wrap(buffer, index, length);
+        routeId = begin.streamId();
+        replyId = factory.supplyReplyId.applyAsLong(begin.streamId());
+        connect = new HttpCacheProxyGroupRequest(factory, this)::onRequestMessage;
+        return connect;
+    }
+
+    private void initiateRequest(
+        String etag,
+        long acceptRouteId,
+        long acceptReplyId)
+    {
+        this.etag = etag;
+        this.acceptRouteId = acceptRouteId;
+        this.acceptReplyId = acceptReplyId;
+        writer.doSignal(acceptRouteId,
+                        acceptReplyId,
+                        factory.supplyTraceId.getAsLong(),
+                        INITIATE_REQUEST_SIGNAL);
     }
 
     private void doSignalCacheEntryAborted(
         long acceptReplyId,
         long acceptRouteId)
     {
+        gotResponse(acceptReplyId);
         writer.doSignal(acceptRouteId,
                         acceptReplyId,
                         factory.supplyTraceId.getAsLong(),
                         CACHE_ENTRY_ABORTED_SIGNAL);
     }
 
+    private void gotResponse(long acceptReplyId)
+    {
+        if (!requestsWithAnswer.containsLong(acceptReplyId))
+        {
+            requestsWithAnswer.pushLong(acceptReplyId);
+        }
+    }
```

Comment:
Should `requestsWithAnswer` be a `LongHashSet` instead of `LongArrayList`, if we only want each value to be present once?

Reply:
Now that this is a set, we no longer need to guard with if-contains check, right?